src/main/java/io/cryostat/graphql/ActiveRecordings.java

```diff
@@ -119,6 +120,31 @@ public RecordingOptions asOptions() {
         }
     }
 
+    @Description(
+            "Updates the metadata labels for an existing Flight Recorder recording specified by the"
+                    + " recording ID.")
+    public Uni<ActiveRecording> doPutMetadata(
+            @Source ActiveRecording recording, MetadataInput metadataInput) {
+        return Uni.createFrom()
+                .item(
+                        () -> {
+                            Map<String, String> labels = metadataInput.getLabels();
+                            return recordingHelper.updateRecordingMetadata(recording.id, labels);
+                        });
+    }
+
+    public static class MetadataInput {
```

Comment:
This part of the schema is similar to https://github.com/cryostatio/cryostat3/pull/294/files#diff-fe2f999e422b5821ed225bca9e8dd197c9d208c0c0937aa0b6c8f26af3284d16R109 - in that one, it's an input that (should - I don't know if I've actually tested it yet) allows the user to set the metadata on the recording when starting it. I think the API schema for setting the metadata at start vs updating it later should be the same, both so that it's simpler for clients and also so that it's a unified data model on the backend.

----

And just to clarify what the new version of this schema should look like to begin with -

In 2.4, the schema looks like this:

````graphql
type ActiveRecording {
    doPutMetadata(metadata: Object): ActiveRecording
}
````

and the implementation:

```java

    @Override
    public HyperlinkedSerializableRecordingDescriptor getAuthenticated(
            DataFetchingEnvironment environment) throws Exception {
        GraphRecordingDescriptor source = environment.getSource();
        ServiceRef target = source.target;
        String uri = target.getServiceUri().toString();
        String recordingName = source.getName();
        Map<String, Object> settings = environment.getArgument("metadata");
        Map<String, String> labels = new HashMap<>();

        if (settings.containsKey("labels")) {
            List<InputRecordingLabel> inputLabels =
                    gson.fromJson(
                            settings.get("labels").toString(),
                            new TypeToken<List<InputRecordingLabel>>() {}.getType());
            for (InputRecordingLabel l : inputLabels) {
                labels.put(l.getKey(), l.getValue());
            }
        }
```

so this means that in 2.4, the schema specifies clients should send a request that looks like this partial (see [here](https://github.com/cryostatio/cryostat/blob/main/src/test/java/itest/GraphQLIT.java#L276) for a real example ):

```graphql
query {
  targetNodes(filter: something) {
    recordings {
      active(filter: somethingElse) {
        doPutMetadata(metadata: {
          labels: [
            {
              key: "foo",
              value: "bar"
            }
          ]
        }
      }
    }
  }
}
```

Since the `recordings: { active }` key-value schema has also been updated, I think this should look like this in the new schema:

```graphql
query {
  targetNodes(filter: something) {
    activeRcordings(filter: somethingElse) {
      doPutMetadata(metadata: {
        labels: [
          {
            key: "foo",
            value: "bar"
          }
        ]
      }
    }
  }
}
```

and there should also be an equivalent `archivedRecordings { doPutMetadata }` that works the same way.

(luckily, the old schema was already using the same key-value form for the labels map, so that doesn't need to be adjusted)

Reply:
tl;dr I think this `MetadataInput` can be removed or refactored to use the already existing `Metadata` type, but it might take more refactoring to get it there. But the schema shouldn't change much and should match what I wrote above.